### PR TITLE
ENT-3289: automatic registration on public cloud

### PR DESCRIPTION
**Pull request by jirihnidek**
* When auto_registration = 1 in rhsm.conf, then rhsmcertd will
  try to do automatic registration on VMs running on public
  clouds. Currently three public clouds are supported:
  - AWS
  - Azure
  - GCP
* Another option auto_registration_interval can be set in rhsm.conf.
  This option sets interval between auto-registratin attempts.
  Default delay is 3600 seconds (one hour).
* The rhsmcertd tries to do only three attempts from the start
  of rhmcertd
* How does it all work?
  - When auto_registration is set to 1, then rhscertd triggers
    rhsmcertd_worker with --auto-register option (three times).
  - When rhsmcert_worker is started with --auto-register, then
    it tries to detect if VM is running on public cloud.
  - When known public cloud is detected, then worker tries to
    collect public cloud metadata and signature
  - When metadata and signature is collected, then it is
    encoded using base64 and these values are set into
    json document.
  - This json document is sent in PUT http request to:
    /cloud/authorise
  - When cloud registration adapter was able to find "cloud ID"
    in provided metadata, then candlepin returns JWT token
    containing organization ID
  - When JTW token is returned, then it is set in http header
    in register request and system is registered as usual
    - No auto-attach
* Added some unit tests
* TODO:
  - Update man pages
  - Update bash complete scripts?

### CODE REVIEW

**Comment by jirihnidek** [comment]
Blocked by #2381

File: test/rhsmlib_test/test_cloud.py [hunk 1]
```diff
@@ -393,3 +916,67 @@ def test_conclict_in_strong_signs(self):
...
+        metadata = base64.b64decode(b64_metadata).decode('utf-8')
+        self.assertEqual(metadata, AWS_METADATA)
+        # Test signature
+        self.assertTrue('signature' in cloud_info)
+        b64_signature = cloud_info['signature']
+        signature = base64.b64decode(b64_signature).decode('utf-8')
+        self.assertEqual(
+            signature,
+            '-----BEGIN PKCS7-----\n' + AWS_SIGNATURE + '\n-----END PKCS7-----'
+        )
```
Comment (cnsnyder) [inline test/rhsmlib_test/test_cloud.py:982]: We'll have to double check but I think this will fail the style checks. I think we should add a line here.

**Review by cnsnyder — CHANGES_REQUESTED** [review]
Overall great work so far!
Aside from the comments below, is there somewhere we could log which cloud providers we have detected and will try as well as which of them have been successful in returning metadata?

Please also see the inline comments / requests.

Thank you very much for your work so far!

**Comment by jirihnidek** [comment]
> Overall great work so far!
> Aside from the comments below, is there somewhere we could log which cloud providers we have detected and will try as well as which of them have been successful in returning metadata?

Good point. Added more `log.info()` and `log.warning()` calls.

> Please also see the inline comments / requests.
> 
> Thank you very much for your work so far!

File: man/rhsm.conf.5 [hunk 2]
```diff
@@ -248,6 +248,16 @@ were set to 3 minutes, the initial cert check would begin somewhere between 2 mi
 disable
 .RS 4
 set to 1 to disable rhsmcertd operation entirely.
+.RE
+.PP
+auto_registration
+.RS 4
+set to 1 to enabled automatic registration. Automatic registration can work only on VMs running in the public cloud. Currently three public cloud providers are supported: AWS, Azure and GCP. When it is desired to perform automatic registration by rhsmcertd, then it is also necessary to configure mapping of "Cloud ID" to "RHSM organization ID" on https://cloud.redhat.com.
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply man/rhsm.conf.5:255]: Fixed

File: etc-conf/rhsm.conf [hunk 3]
```diff
@@ -94,6 +94,10 @@ autoAttachInterval = 1440
 splay = 1
 # If set to 1, rhsmcertd will not execute.
 disable = 0
+# Set to 1, when rhsmcerd will try to do automatic registration.
+# Setting this option make sense only on machines running on public
+# clouds. Currently only AWS, Azure and GCP are supported
+auto_registration = 0
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply etc-conf/rhsm.conf:100]: Good point!

File: src/rhsmlib/cloud/providers/azure.py [hunk 4]
```diff
@@ -18,7 +18,15 @@
...
 
+import logging
+
+from typing import Union
+
 from rhsmlib.cloud.detector import CloudDetector
+from rhsmlib.cloud.collector import CloudCollector
+
+
+log = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply src/rhsmlib/cloud/providers/azure.py:29]: Good point.

File: src/rhsm/config.py [hunk 5]
```diff
@@ -83,7 +83,8 @@
         'certcheckinterval': '240',
         'autoattachinterval': '1440',
         'splay': '1',
-        'disable': '0'
+        'disable': '0',
+        'auto_registration': 0
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply src/rhsm/config.py:87]: You are right, fixed.

File: src/rhsmlib/cloud/utils.py [hunk 6]
```diff
@@ -18,26 +18,39 @@
...
+    AWSCloudCollector,
+    AzureCloudCollector,
+    GCPCloudCollector
+]
+
 log = logging.getLogger(__name__)
 
 
-def detect_cloud_provider():
+def detect_cloud_provider() -> list:
```
[the remark this answers is not in the document]
Comment (jirihnidek) [reply src/rhsmlib/cloud/utils.py:53]: Very good point, thanks. Fixed.

File: test/rhsmlib_test/test_cloud.py (hunk 1 above)
Comment (jirihnidek) [reply test/rhsmlib_test/test_cloud.py:982]: Fixed.

**Comment by martinpitt** [comment]
I abuse this PR to see whether sub-man's tests work on rhel-8-4 as well. Right now it is only being tested against 8.3, but you should really move ASAP. If that works, I can switch over the cockpit bot's testmap.

**Comment by martinpitt** [comment]
Ah, so that doesn't work at all -- it seems these tests haven't been touched in quite some time, I'll send a PR.

**Review by cnsnyder — APPROVED** [review]

**Pull request merged**